### PR TITLE
`tabnet_pretrain` accepts missings in predictors

### LINKED ISSUE #65: Allow missing values in predictors during pretraining  through a static NA_mask
I can't get comfortable with the idea of preventing NAs
```r
stopifnot("Error: found missing values in the predictor data frame" = sum(is.na(x))==0)
```
Specifically, the `ames` dataset uses a trick of imputing NA with zeros for a lot of numerical predictors that can only biais the model: 
As an example, the "Masonry veneer area" predictor
``` r
suppressPackageStartupMessages(library(tidymodels))
data("ames", package = "modeldata")
qplot(ames$Mas_Vnr_Area)
#> `stat_bin()` using `bins = 30`. Pick better value with `binwidth`.
```

<sup>Created on 2021-10-16 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

No house has a "Masonry veneer area" of surface 0, they actually don't have. Idem for "Pool area"  and a lot more predictors. 
Knowing that we already apply a `random_obfuscation` mask to value in `tabnet_pretrain()`, we could maybe force a binary mask of the predictor NAs to be part of it.

### PR DESCRIPTION
**Pull request by cregouby**
fix #65. and add corresponding vignette.
At the same time
- add `vip::vip()` and `tabnet_explain()` support for `tabnet_pretrain` objects
- improve `random_obfuscator` masking performance
image

### CODE REVIEW

**Comment by cregouby** [comment]
Hello @dfalbel, any chance that you spend some to review this P.R. ?

**Comment by dfalbel** [comment]
@cregouby ! So sorry, i completely missed it. Will review today!
Sorry about that!

**Comment by cregouby** [comment]
Thanks !

**Comment by dfalbel** [comment]
Hey @cregouby ! I couldn't get to it today, but will have better look tomorrow. 
Sorry for the delay.

**Comment by cregouby** [comment]
No worries at all.

**Review by dfalbel — COMMENTED** [review]
Hi @cregouby ! Codewise this looks great! I don't have any comments.

I'm a bit confused by the example though. 
It seems to me that masking the `NA` values would be a good decision when missingness occurs randomly - thus we don't want the model to encode that information. 

In the ames case, I'd say that we in general wont want to discard this information as that will be important for finding the 0 versus >0 relationship. For example has a pool or not.

Do I understand correctly, or maybe I am missing something?

Sorry again for the long time to review your PR.

**Comment by cregouby** [comment]
You are right on the fact that this feature is a given solution for missing at random (MAR) dataset. Maybe it is worth mentioning in the vignette... for Non missing at random (NMAR) , like in Ames, the question is definitively valid and covers multiple topics :
## human perspective
This is related to the information encoded in the variable, and their semantic, nothing related to the modeling aspects
- do we understand that the value `pool surface = 0 m²` is a proxy information for the implicit column "has_pool = FALSE" (I think yes,...)
- how could we understand the value `pool surface = 0.01 m²` in the dataset, as we encode it as a numerical positive value, this should be valid, and could be the result of any regression.
- same for categorical qualitative feature on the pool as `pool_condition = "no_pool"` is also a proxy encoding, not a pool condition...
So that's what I had in mind when I prepared the `ames_missing` dataset.

## modeling quality perspective
Now, is it beneficial for the model performance ? 
- is the model technically capable of forging an internal representation the proxy variable `has_pool` based on the interaction between `pool_quality = "no_pool"` and `pool_surface = 0m²` ? I think answer is yes as it is the aim of the attention block to give room to such feature.
- has the model enough data to do it with ames dataset ? I don't know.
- will the pretrained model be better with explicit NAs or with encoded NAs for the following task of ames price prediction ? I don't know (and I eluded the question in the vignette not doing it further...)
- will the pretrained model be better to capture this interaction with explicit NAs or with encoded NAs  ? Maybe not, as you suggest, but at least variable importance output is more to be trusted with explicit NAs ( as the vignette highlight), and this is an interesting topic (that I just surface for now)

I may have a slot to discuss the fundamental part of it with one of the `missing-data` cran task view author. In the meantime, I don't know what's best....

**Comment by dfalbel** [comment]
@cregouby OK! I'm much more confident that I fully understand the approach for handling missing in the training data.
I agree with you there are many factors to discuss around that theme and there's no perfect solution.

I think for completeness it would be nice to add a paragraph in the motivation section of the vignette describing in a high level how we approach the problem in TabNet and what are the wins and possible drawbacks. That way I feel users will be more confident in relying on TabNet's missing data handling.

What do you think?

**Comment by cregouby** [comment]
It sounds perfect for me !

**Comment by cregouby** [comment]
Hello @dfalbel,
Finally, presence or absence of missing data in the vip plot is not reproducible / was by chance / has such variability in between  training occurrences. So I prefer to remove the vignette ( that was at the end more like a blog article ) 
We may revive it later on (when/if I can manage NAs in downstream tasks )

**Comment by dfalbel** [comment]
@cregouby Sounds great! Feel free to merge this PR whenever you want.
